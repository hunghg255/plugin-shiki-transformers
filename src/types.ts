import type { HtmlRendererOptions } from 'shiki'

export type TLineOptions = NonNullable<HtmlRendererOptions['lineOptions']>
export interface ITransformerResult {
	code: string
  lineOptions: TLineOptions
}

export type TPostTransformerResult = string | undefined

export interface ITransformerOptions {
	code: string
	lang: string
}

export type TPreTransformer = (options: ITransformerOptions) => ITransformerResult
export type TPostTransformerHandler = (options: ITransformerOptions) => TPostTransformerResult

export interface IRangeTransformerOptions {
	tagRegExp?: RegExp
}

export interface ITransformer {
	name: string
	preTransformer?: TPreTransformer
	postTransformer?: TPostTransformerHandler
}
