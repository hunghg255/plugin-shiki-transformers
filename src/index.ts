export * from './transformers'
export * from './utils'
export * from './types'

export { defineTransformer, transformer, postTransformer } from './transformer'
export { getHighlighter } from './highlighter'
