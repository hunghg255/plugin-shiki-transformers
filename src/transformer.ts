import type { ITransformer, ITransformerResult } from './types'

/**
 * Defines a transformer.
 */
export function defineTransformer(transformer: ITransformer): ITransformer {
  return transformer
}

/**
 * Transforms code through the given transformer.
 */
export function transformer(transformers: ITransformer[], code: string, lang: string) {
  return transformers.reduce((options, transformer) => {
    const { code, lineOptions } = transformer.handler?.({
      code: options.code,
      lang,
    }) ?? options

    return {
      code,
      lineOptions: [
        ...options.lineOptions,
        ...lineOptions,
      ],
    }
  }, {
    code,
    lineOptions: [],
  } as ITransformerResult)
}

/**
 * Transforms final code through the given Transformers.
 */
export function postTransformer(transformers: ITransformer[], code: string, lang: string) {
  return transformers.reduce((code, transformer) => transformer.postTransformer?.({
    code,
    lang,
  }) ?? code, code)
}
