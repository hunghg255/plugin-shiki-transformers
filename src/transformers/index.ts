export * from './diff'
export * from './focus'
export * from './highlight'
export * from './highlight-error-level'
export * from './line-number'
