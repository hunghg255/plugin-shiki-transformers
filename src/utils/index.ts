export * from './add-class'
export * from './create-range-transformer'
