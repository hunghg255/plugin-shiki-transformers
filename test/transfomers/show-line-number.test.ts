import { expect, it } from 'vitest'
import {  createTransformerLineNumber, getHighlighter } from '../../src'

async function testTransformerLineNumber(snippet: string) {


	const highlighter = await getHighlighter({
		theme: 'nord',
		transformers: [
			createTransformerLineNumber(),
		],
	})

	expect(highlighter.codeToHtml(snippet, { lang: 'javascript' })).toMatchSnapshot()
}

it('generates show lines on the same line as the tag', async() => {
	const snippet = `
console.log('owo'); // [!code hl]
console.log('owo'); // [!code hl]
	`

	await testTransformerLineNumber(snippet)
})
