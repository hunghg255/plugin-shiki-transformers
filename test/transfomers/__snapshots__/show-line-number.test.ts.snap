// Vitest Snapshot v1

exports[`generates show lines on the same line as the tag 1`] = `
"<pre class=\\"shiki nord\\" style=\\"background-color: #2e3440ff\\" tabindex=\\"0\\"><code><span class=\\"line line-number\\"></span>
<span class=\\"line line-number\\"><span style=\\"color: #D8DEE9\\">console</span><span style=\\"color: #ECEFF4\\">.</span><span style=\\"color: #88C0D0\\">log</span><span style=\\"color: #D8DEE9FF\\">(</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #A3BE8C\\">owo</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #D8DEE9FF\\">)</span><span style=\\"color: #81A1C1\\">;</span><span style=\\"color: #D8DEE9FF\\"> </span><span style=\\"color: #616E88\\">// [!code hl]</span></span>
<span class=\\"line\\"><span style=\\"color: #D8DEE9\\">console</span><span style=\\"color: #ECEFF4\\">.</span><span style=\\"color: #88C0D0\\">log</span><span style=\\"color: #D8DEE9FF\\">(</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #A3BE8C\\">owo</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #D8DEE9FF\\">)</span><span style=\\"color: #81A1C1\\">;</span><span style=\\"color: #D8DEE9FF\\"> </span><span style=\\"color: #616E88\\">// [!code hl]</span></span>
<span class=\\"line\\"><span style=\\"color: #D8DEE9FF\\">	</span></span></code></pre>"
`;
